homeassistant/components/proxmoxve/__init__.py

```diff
@@ -91,100 +94,132 @@
-    hass.data[DOMAIN][COORDINATORS] = coordinators
 
-    # Create a coordinator for each vm/container
-    for host_config in config[DOMAIN]:
-        host_name = host_config["host"]
-        coordinators[host_name] = {}
+    proxmox = await hass.async_add_executor_job(proxmox_client.get_api_client)
 
-        proxmox_client = hass.data[PROXMOX_CLIENTS][host_name]
+    node = entry_data[CONF_NODE]
+    node_coordinators = coordinators[node] = {}
 
-        # Skip invalid hosts
-        if proxmox_client is None:
-            continue
+    # Proxmox instance info
+    coordinator = create_coordinator_container_vm(
+        hass, proxmox, entry_data[CONF_HOST], None, None, ProxmoxType.Proxmox
+    )
 
-        proxmox = proxmox_client.get_api_client()
+    # Fetch initial data
+    await coordinator.async_refresh()
 
-        for node_config in host_config["nodes"]:
-            node_name = node_config["node"]
-            node_coordinators = coordinators[host_name][node_name] = {}
+    node_coordinators[ProxmoxType.Proxmox] = coordinator
 
-            for vm_id in node_config["vms"]:
-                coordinator = create_coordinator_container_vm(
-                    hass, proxmox, host_name, node_name, vm_id, TYPE_VM
-                )
+    # Node info
+    coordinator = create_coordinator_container_vm(
+        hass, proxmox, entry_data[CONF_HOST], node, None, ProxmoxType.Node
+    )
 
-                # Fetch initial data
-                await coordinator.async_refresh()
+    # Fetch initial data
+    await coordinator.async_refresh()
 
-                node_coordinators[vm_id] = coordinator
+    node_coordinators[ProxmoxType.Node] = coordinator
 
-            for container_id in node_config["containers"]:
-                coordinator = create_coordinator_container_vm(
-                    hass, proxmox, host_name, node_name, container_id, TYPE_CONTAINER
-                )
+    # QEMU info
+    for vm_id in entry_data[CONF_QEMU]:
+        coordinator = create_coordinator_container_vm(
+            hass,
+            proxmox,
+            entry_data[CONF_HOST],
+            node,
+            vm_id,
+            ProxmoxType.QEMU,
+        )
 
-                # Fetch initial data
-                await coordinator.async_refresh()
+        # Fetch initial data
+        await coordinator.async_refresh()
 
-                node_coordinators[container_id] = coordinator
+        node_coordinators[vm_id] = coordinator
 
-    for component in PLATFORMS:
-        await hass.async_create_task(
-            async_load_platform(hass, component, DOMAIN, {"config": config}, config)
+    # LXC info
+    for container_id in entry_data[CONF_LXC]:
+        coordinator = create_coordinator_container_vm(
+            hass,
+            proxmox,
+            entry_data[CONF_HOST],
+            node,
+            container_id,
+            ProxmoxType.LXC,
+        )
+
+        # Fetch initial data
+        await coordinator.async_refresh()
```

Comment:
Since you have multiple (4?) coordinators that need to refresh, I would bundle them together as `asyncio` tasks, then use `asyncio.gather` to run them simultaneously. If you make sure that each uses `async_config_entry_first_refresh`, you can know that `ConfigEntryNotReady` will get raised if any fails.

Reply:
I solved almost all the revisions, but this one I will need to study a little. I'm just an enthusiast, so my knowledge is a little limited.

Do you have any examples that could help?

Reply:
You bet; check out what I did for RainMachine: https://github.com/home-assistant/core/blob/01f1629ac0763a1634a3dc001dd84a5ed92a461f/homeassistant/components/rainmachine/__init__.py#L252-L272

Reply:
I adapted something, but I don't know if it was expected.

Please see if you agree.

https://github.com/home-assistant/core/blob/333cb8e2913eda5bd57f60a0779c9eb5b483f2de/homeassistant/components/proxmoxve/__init__.py#L242-L274

Thanks!